src/ArrayTemplates.hs

```diff
@@ -61,23 +75,42 @@ templateEFilter = defineTypeParameterizedTemplate templateCreator path t docs
           t
           (const (toTemplate "Array $NAME(Lambda *predicate, Array a)")) -- Lambda used to be $(Fn [(Ref a)] Bool)
           ( \(FuncTy [RefTy (FuncTy [RefTy insideTy _] BoolTy _) _, _] _ _) ->
-              toTemplate $
-                unlines $
-                  let deleter = insideArrayDeletion typeEnv env insideTy
-                   in [ "$DECL { ",
-                        "    int insertIndex = 0;",
-                        "    for(int i = 0; i < a.len; ++i) {",
-                        "        if(" ++ templateCodeForCallingLambda "(*predicate)" fTy [elt] ++ ") {",
-                        "            ((($a*)a.data)[insertIndex++]) = (($a*)a.data)[i];",
-                        "        } else {",
-                        "        " ++ deleter "i",
-                        "        }",
-                        "    }",
-                        "    a.len = insertIndex;",
-                        templateShrinkCheck "a",
-                        "    return a;",
-                        "}"
-                      ]
+              multilineTemplate
+                ( let deleter = insideArrayDeletion typeEnv env insideTy
+                   in ( case insideTy of
+                          UnitTy ->
+                            [ "$DECL { ",
+                              "    int insertIndex = 0;",
+                              "    for(int i = 0; i < a.len; ++i) {",
+                              "        if(" ++ templateCodeForCallingLambda "(*predicate)" fTy [elt] ++ ") {",
+                              "            /* ignore () member; just increment length. */",
```

Comment:
This template really looks like it should be possible to write without duplicating the complete body, but rather extract the inner loop part (much like in endoMap above), no?

Reply:
Good catch, updated!

Reply:
cb7800e